Add Iceberg Extension Documentation (#314)
kuzudb/kuzu-docs: src/content/docs/extensions/iceberg.mdx

---
title: "iceberg extension"
---

import { Tabs, TabItem } from '@astrojs/starlight/components';

## Usage

The `iceberg` extension adds support for `iceberg` objects, currently including scanning from [Apache Iceberg format](https://iceberg.apache.org/).

The iceberg functionality is not available by default, so you would first need to install the iceberg
extension by running the following commands:

```sql
INSTALL iceberg;
LOAD EXTENSION iceberg;
```

### Example dataset

> To test the examples, download the [iceberg_data.zip](TODO: upload iceberg sample data for download) file and unzip it.

Author: we decided to put the download link in doc page

## Usage

### Common Parameters

```
┌───────────────────────────────┬───────────┬──────────────┬──────────────────────────────────────────────────────────────────────────────┐
│ Parameter                     │ Type      │ Default      │ Description                                                                  │
├───────────────────────────────┼───────────┼──────────────┼──────────────────────────────────────────────────────────────────────────────┤

> │ allow_moved_paths             │ BOOLEAN   │ false        │ Allows scanning Iceberg tables that are moved                                │

Author: illustrated

│ metadata_compression_codec    │ STRING    │ ''           │ Treats metadata files as when set to 'gzip'                                  │

> │ version                       │ STRING    │ '?'          │ Provides an explicit version string, hint file or guessing                   │

Reviewer: What do you mean "hint file or guessing"?

Author: removed this unclear statement

> │ version_name_format           │ STRING    │ 'v%s%s.metadata.json,%s%s.metadata.json' │ Controls how versions are converted to metadata file names │

Reviewer: I think the Options table is too brief to explain the idea behind these options, you should have a separate paragraph on explaining each option more clearly. 
Each option should come with an example query to demo how they are set/used.

When I go over this as someone not an expert on iceberg, I got confused:
1. What does moved paths mean? and how the behaviour would be when `allow_moved_paths` is set to `true` or `false`, respectively?
2. What compression codec do we support for `metadata_compression_codec`? Default value `''` means uncompressed?
3. What are the rules for version name format?

Author: restructured this

```

### Scan the Iceberg file

```cypher
LOAD FROM 'kuzu/extension/iceberg/test/iceberg_tables/person_table' (file_format='iceberg', allow_moved_paths=true) RETURN *;
```
```
┌─────────┬───────┬──────────────┬─────────────────────────────┐
│ Name    │ Age   │ Balance      │ Date of Birth               │
├─────────┼───────┼──────────────┼─────────────────────────────┤
│ Alice   │ 25    │ 1000.500000  │ 1999-01-15 05:00:00+00      │
│ Bob     │ 30    │ 1500.750000  │ 1994-06-25 04:00:00+00      │
│ Charlie │ 35    │ 2000.000000  │ 1988-12-05 05:00:00+00      │
│ Diana   │ 40    │ 2500.300000  │ 1983-07-20 04:00:00+00      │
│ Ethan   │ 28    │ 1800.600000  │ 1996-03-10 05:00:00+00      │
└─────────┴───────┴──────────────┴─────────────────────────────┘

```

:::note[Note]
The `allow_moved_paths` option ensures that some path resolution is performed, which allows scanning Iceberg tables that are moved.
:::

This extension can be paired with the httpfs extension to access Iceberg tables in object stores such as S3.

```
LOAD FROM 's3://kuzu-test/iceberg' (file_format='iceberg', allow_moved_paths=true) RETURN *;
```

### Access Iceberg Metadata

```cypher

> CALL ICEBERG_METADATA('kuzu/extension/iceberg/test/iceberg_tables/lineitem_iceberg', allow_moved_paths := true) return *;

Author: The lineitem_iceberg is mainly used to illustrate the usage of option `version`, I didn't find a way to generate iceberg table with different version, let me work on that

> CALL ICEBERG_METADATA('kuzu/extension/iceberg/test/iceberg_tables/lineitem_iceberg', allow_moved_paths := true) return *;

Author: We now can generate the iceberg table with different version, but different compression code is not figured out

```

```
┌──────────────────────────┬──────────────────────────┬──────────────────┬─────────┬──────────┬──────────────────────────┬─────────────┬──────────────┐
│ manifest_path            │ manifest_sequence_number │ manifest_content │ status  │ content  │ file_path                │ file_format │ record_count │
│ STRING                   │ INT64                    │ STRING           │ STRING  │ STRING   │ STRING                   │ STRING      │ INT64        │
├──────────────────────────┼──────────────────────────┼──────────────────┼─────────┼──────────┼──────────────────────────┼─────────────┼──────────────┤
│ lineitem_iceberg/meta... │ 2                        │ DATA             │ ADDED   │ EXISTING │ lineitem_iceberg/data... │ PARQUET     │ 51793        │
│ lineitem_iceberg/meta... │ 2                        │ DATA             │ DELETED │ EXISTING │ lineitem_iceberg/data... │ PARQUET     │ 60175        │
└──────────────────────────┴──────────────────────────┴──────────────────┴─────────┴──────────┴──────────────────────────┴─────────────┴──────────────┘

```

:::note[Note]

> Be sure to use `:=` operator to set the variable in `CALL` function, as `=` operator in `CALL` function has a different meaning of comparison.

Reviewer: Did we document this somewhere else along with CALL? I guess this is a common grammar rule not specific to iceberg but all CALL functions. @acquamarin Can you double check?

Author: keep this as non resolved as a reminder

Reviewer 2: I'll keep a track of this separately.

:::

> ### Visualizing Snapshots

Reviewer: What do you mean "Visualize" here?

Reviewer 2: I think we should just phrase this as "List snapshots"

Author: changed to list

extension by running the following commands:

```sql
INSTALL iceberg;
LOAD EXTENSION iceberg;
```

### Example dataset

To test the examples, download the [iceberg_data.zip](TODO: upload iceberg sample data for download) file and unzip it.

### Common Functions

#### Scan the Iceberg table
```cypher
LOAD FROM 'iceberg_table_path' (file_format='iceberg') RETURN *;
```
#### Copy the Iceberg table into a node table
```cypher
COPY table_name FROM 'iceberg_table_path' (file_format='iceberg');
```
#### Access the Iceberg Metadata
```cypher
CALL ICEBERG_METADATA('iceberg_table_path') RETURN *;
```
#### List the Iceberg Snapshots
```cypher
CALL ICEBERG_SNAPSHOTS('iceberg_table_path') RETURN *;
```

### Optional parameters

| Parameter                     | Type      | Default      | Description                                                                     |
|-------------------------------|-----------|--------------|---------------------------------------------------------------------------------|
| allow_moved_paths             | BOOLEAN   | false        | Allows scanning Iceberg tables that are not located in their original directory |

> | metadata_compression_codec    | STRING    | ''           | Specifies the compression code used for the metadata files                      |

Reviewer: @SterlingT3485 I think you mentioned the only supported compression codec is `gzip`? Should we state that here?

Author: Yes, let me add that in the table

Author: Added

## Usage

### Common Functions

#### Scan the Iceberg table
```cypher
LOAD FROM 'iceberg_table_path' (file_format='iceberg') RETURN *;
```
#### Copy the Iceberg table into a node table
```cypher
COPY table_name FROM 'iceberg_table_path' (file_format='iceberg');
```
#### Access the Iceberg Metadata
```cypher
CALL ICEBERG_METADATA('iceberg_table_path') RETURN *;
```
#### List the Iceberg Snapshots
```cypher
CALL ICEBERG_SNAPSHOTS('iceberg_table_path') RETURN *;
```

> ### Common Options

Reviewer 2: ```suggestion
### Optional parameters
```

│ allow_moved_paths             │ BOOLEAN   │ false        │ Allows scanning Iceberg tables that are not located in their original directory │
│ metadata_compression_codec    │ STRING    │ ''           │ Specifies the compression code used for the metadata files                   │
│ version                       │ STRING    │ '?'          │ Provides an explicit Iceberg version number, if not provided, the Iceberg version number would be determined from `version-hint.txt`│
│ version_name_format           │ STRING    │ 'v%s%s.metadata.json,%s%s.metadata.json' │ Provides the regular expression to find the correct metadata data file │
└───────────────────────────────┴───────────┴──────────────┴──────────────────────────────────────────────────────────────────────────────┘
```

### Scan the Iceberg table

`LOAD FROM` is a Cypher query that scans a file or object element by element, but doesn’t actually
move the data into a Kùzu table.

To scan an iceberg table, you can do the following:

```cypher
LOAD FROM 'kuzu/extension/iceberg/test/iceberg_tables/person_table' (file_format='iceberg', allow_moved_paths=true) RETURN *;
```

> Note: The `file_format` parameter is used to explicitly specify the file format of the given file instead of letting kuzu sniff the file format at runtime. When scanning from the Iceberg table, 

Reviewer 2: ```suggestion
Note: The `file_format` parameter is used to explicitly specify the file format of the given file instead of letting Kùzu autodetect the file format at runtime. When scanning from the Iceberg table, 
```

```

### Limitations

> Writing (i.e., exporting to) Iceberg files is currently not supported.

Reviewer 2: ```suggestion
 - Writing (i.e., exporting to) Iceberg files is currently not supported.
```

> The current version of Iceberg Extension doesn't support structs inside iceberg table.

Reviewer 2: ```suggestion
- The current version of Iceberg Extension doesn't support structs inside iceberg table.
```
